### PR TITLE
feat(components/colorpicker): add colorpicker test harness 

### PR DESCRIPTION
**Pull request by Blackbaud-SandhyaRajasabeson**
[AB#2195390](https://dev.azure.com/blackbaud/Products/_boards/board/t/SKY%20UX%20Program/Stories/?workitem=2195390)

### CODE REVIEW

**Comment by blackbaud-sky-build-user** [comment]
[Storybook preview](https://blackbaud.github.io/skyux-pr-preview/2768/storybook/)

Component Storybooks:

- _(no component storybooks affected in this pr)_ 

Apps:

- [code-examples](https://blackbaud.github.io/skyux-pr-preview/2768/code-examples/)
- [integration](https://blackbaud.github.io/skyux-pr-preview/2768/integration/)
- [playground](https://blackbaud.github.io/skyux-pr-preview/2768/playground/)

File: libs/components/help-inline/testing/src/help-inline/help-inline-harness.ts [hunk 1]
```diff
@@ -105,9 +104,7 @@ export class SkyHelpInlineHarness extends SkyComponentHarness {
   /**
    * Gets the help popover content.
    */
-  public async getPopoverContent(): Promise<
-    TemplateRef<unknown> | string | undefined
-  > {
+  public async getPopoverContent(): Promise<string | undefined> {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/help-inline/testing/src/help-inline/help-inline-harness.ts]: We should revert this change since it would be considered "breaking".

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  }
+
+  /**
+   * Whether the colorpicker component is stacked.
+   */
+  public async isStacked(): Promise<boolean> {
+    return (await this.host()).hasClass('sky-margin-stacked-lg');
+  }
+
+  private async getOverlayId(): Promise<string | null> {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts]: `#getOverlayId`

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    const overlayId = await this.getOverlayId();
+
+    if (!overlayId) {
+      throw new Error(
+        'Unable to get colorpicker dropdown because dropdown is closed',
+      );
+    }
+
+    return await this.#documentRootLocator.locatorFor(
+      SkyColorpickerDropdownHarness.with({ id: overlayId }),
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts]: You probably don't need to create a new `id` filter for this. See how the dropdown harness uses the `selector` filter: https://github.com/blackbaud/skyux/blob/main/libs/components/popovers/testing/src/dropdown/harness/dropdown-harness.ts#L87-L89

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+  private async getId(): Promise<string | null> {
+    return (await this.host()).getAttribute('id');
+  }
+
+  private async getRedInput(): Promise<TestElement> {
+    return (await this.#getInputs())[1];
+  }
+
+  private async getSwatchButtons(): Promise<TestElement[]> {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts]: ☝️ All the "private" methods should start with `#`.

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.filters.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { BaseHarnessFilters } from '@angular/cdk/testing';
+
+/**
+ * A set of criteria that can be used to filter a list of `SkyColorpickerDropdownHarness` instances.
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface SkyColorpickerDropdownHarnessFilters
+  extends BaseHarnessFilters {
+  id?: string;
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.filters.ts]: Probably not needed (see my comment below about using `selector`).

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+  /**
+   * Enters a value into the alpha input box.
+   * @param value A decimal value from 0-1.
+   */
+  public async setAlphaValue(value: string): Promise<void> {
+    const input = await this.getAlphaInput();
+
+    input.setInputValue(value);
+    input.dispatchEvent('input');
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts]: This operation is copied in multiple methods. What do you think about creating a shared private method to handle all cases?
```
const input = await this.getAlphaInput();
await this.#setInputValue(input, value);
```

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+      }
+    }
+  }
+
+  /**
+   * Enters a value into the alpha input box.
+   * @param value A decimal value from 0-1.
+   */
+  public async setAlphaValue(value: string): Promise<void> {
+    const input = await this.getAlphaInput();
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts]: This should throw a helpful error if the alpha button cannot be found (since it's enabled by the `allowTransparency` flag).

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+   */
+  public async clickCancelButton(): Promise<void> {
+    return (await this.#getCancelButton()).click();
+  }
+
+  /**
+   * Clicks a specified swatch in the color preset section.
+   * @param swatchHex Hex code of the swatch to click.
+   */
+  public async clickSwatch(swatchHex: string): Promise<void> {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts]: Let's name this `clickPresetColorSwatch` to match the name used in the design guidelines anatomy section. This will also protect us if we create a different type of swatch in the future.

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+  );
+  #resetButtonModern = this.locatorForOptional(
+    'button.sky-colorpicker-reset-button-modern',
+  );
+
+  async #getFormErrors(): Promise<SkyFormErrorsHarness> {
+    return await this.locatorFor(SkyFormErrorsHarness)();
+  }
+
+  async #getHelpInline(): Promise<SkyHelpInlineHarness> {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts]: Private methods should go beneath public ones.

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    return await this.#documentRootLocator.locatorFor(
+      SkyColorpickerDropdownHarness.with({ id: overlayId }),
+    )();
+  }
+
+  /**
+   * Gets the `SkyIconHarness` for the colorpicker icon.
+   * @internal
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts:107]: Do we even need this method since it's exposing an internal detail of the colorpicker?

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+      return await defaultButton?.click();
+    } else if (modernButton) {
+      return await modernButton?.click();
+    }
+  }
+
+  /**
+   * Gets the colorpicker component's aria label.
+   */
+  public async getAriaLabel(): Promise<string | null> {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts:76]: There are 11 `aria-label` attributes being set on the colorpicker. The JSDocs say "colorpicker component's aria label" but it's not clear what that means. Maybe make this method more specific to which `aria-label` you're retrieving and update the JSDocs accordingly?

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+   * Gets the colorpicker component's label text.
+   */
+  public async getLabelText(): Promise<string> {
+    return (await (await this.#getLabel()).text()).trim();
+  }
+
+  /**
+   * Gets the colorpicker button's title.
+   */
+  public async getTitle(): Promise<string | null> {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts]: This is an internal detail of the component and probably shouldn't be leaked to the consumer.

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+   * Whether the required error has fired.
+   */
+  public async hasRequiredError(): Promise<boolean> {
+    return (await this.#getFormErrors())?.hasError('required');
+  }
+
+  /**
+   * Whether the colorpicker component's label is hidden.
+   */
+  public async isLabelHidden(): Promise<boolean> {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts]: We name this `getLabelHidden` in the other harnesses.

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    }
+    if (defaultButton) {
+      return await defaultButton?.click();
+    } else if (modernButton) {
+      return await modernButton?.click();
+    }
+  }
+
+  /**
+   * Gets the colorpicker component's aria label.
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts]: For all your JSDocs comments: when talking about ARIA, it should be all-caps; if you're referencing a specific attribute, put it in ticks so it's clear to the consumer: \`aria-label\`.

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+   */
+  public async getAriaLabel(): Promise<string | null> {
+    return (await this.#getButton()).getAttribute('aria-label');
+  }
+
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts:79]: Do we have a method to retrieve the `aria-labelledby` value? There's a `labelledBy` input on the colorpicker, so we should probably add one. Maybe take a look at the help inline harness for a similar implementation:
https://github.com/blackbaud/skyux/blob/main/libs/components/help-inline/testing/src/help-inline/help-inline-harness.ts#L80-L90

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    if (!defaultButton && !modernButton) {
+      throw new Error('No reset button found.');
+    }
+    if (defaultButton) {
+      return await defaultButton?.click();
+    } else if (modernButton) {
+      return await modernButton?.click();
+    }
+  }
+
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts:72]: There's an input for `showResetButton`, but there's not an easy way to check that it exists or not (without catching an error when clicking on it). Maybe add a `hasResetButton(): Promise<boolean>` method?

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+   * Enters a value into the alpha input box.
+   * @param value A decimal value from 0-1.
+   */
+  public async setAlphaValue(value: string): Promise<void> {
+    const input = await this.getAlphaInput();
+
+    input.setInputValue(value);
+    input.dispatchEvent('input');
+  }
+
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts:95]: I'm not seeing methods to capture `outputFormat`, `presetColors`, `alphaChannel`, or `allowTransparency` on the `colorpicker-input.directive`.

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts (hunk 10 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts:107]: The only place we use this is in text editor and could be useful for our internal testing purposes, that's why i added it. Also with the internal tag consumers wouldn't see this in our docs and so wouldnt be using it right?

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts (hunk 11 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts:76]: went off of what the button that triggers opening the colorpicker dropdown is called in the anatomy docs

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts (hunk 17 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts:95]: i can add functions for allowTransparency and one that returns an array for presetColors but `outputFormat` and `alphaChannel` does not affect the dom for colorpicker

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+      if ((await swatch.getAttribute('aria-label')) === ariaLabel) {
+        return await swatch.click();
+      }
+    }
+  }
+
+  /**
+   * Gets an array of the hex codes of the preset swatches.
+   */
+  public async getPresetSwatches(): Promise<(string | undefined)[]> {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts]: Small, but we have `clickPresetColorSwatch` above; could we call this method `getPresetColorSwatches`?

Comment (Blackbaud-SteveBrush) [reply libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts]: The return type is an array of `string` or `undefined`. So the result could possibly be `['#fff', undefined]`? Can we ensure only defined values get returned? `Promise<string[]>`

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+   * Gets the colorpicker button's `aria-label`.
+   */
+  public async getAriaLabel(): Promise<string | null> {
+    return (await this.#getButton()).getAttribute('aria-label');
+  }
+
+  /**
+   * Gets the colorpicker button's `aria-labelledBy`
+   */
+  public async getAriaLabelledBy(): Promise<string | null> {
```
Comment (Blackbaud-SteveBrush) [inline libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts]: The attribute has a lowercase "b" for "by" (not great, but it is what it is). We can mimic what we're doing in the other harnesses by calling this `getAriaLabelledby` (the JSDoc also needs to use `aria-labelledby`).

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts (hunk 10 above)
Comment (Blackbaud-SteveBrush) [reply libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts:107]: It won't be documented, but it doesn't mean consumers won't use it. I'm inclined to remove this method until it's absolutely needed.

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts (hunk 19 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts]: I was trying to have it match what we call the input in our docs

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts (hunk 10 above)
Comment (Blackbaud-SandhyaRajasabeson) [reply libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts:107]: mmmmm id be down to comment it out just so its already done and ready for when we do text editor harness

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts (hunk 19 above)
Comment (Blackbaud-SteveBrush) [reply libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts]: The input is named `labelledBy` (does not include "aria"). This method serves to return the value of the native attribute `aria-labelledby`, so we need to rename it accordingly (similar to how we name it in our other harnesses).

**Checkbox, as an example:**
`labelledBy` input: https://github.com/blackbaud/skyux/blob/main/libs/components/forms/src/lib/modules/checkbox/checkbox.component.ts#L85
`getAriaLabelledby` method on the harness: https://github.com/blackbaud/skyux/blob/main/libs/components/forms/testing/src/checkbox/checkbox-harness.ts#L78-L80

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts (hunk 10 above)
Comment (Blackbaud-SteveBrush) [reply libs/components/colorpicker/testing/src/colorpicker/colorpicker-harness.ts:107]: I guess that's fair. Keep it.

File: libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts (hunk 18 above)
Comment (Blackbaud-SteveBrush) [reply libs/components/colorpicker/testing/src/colorpicker/colorpicker-dropdown-harness.ts]: Thanks for fixing the return type of this method. The first question still applies, however.

**Review by Blackbaud-SteveBrush — APPROVED** [review]
